frontend/src/lib/rest/RestRequests.ts

```diff
@@ -0,0 +1,83 @@
+import { showNotification } from 'modules/notifications/notifications';
+
+const username = 'admin';
+const password = 'password';
+const backendUrl = process.env.REACT_APP_BACKEND_HOST;
+const DEFAULT_TIMEOUT = 10000;
+
+export enum Method {
+  GET = 'GET',
+  POST = 'POST',
+  PUT = 'PUT',
+  DELETE = 'DELETE',
+}
+
+const getHeaders = (): Headers => {
+  return new Headers({
+    Authorization: `Basic ${btoa(`${username}:${password}`)})`,
+    'Content-Type': 'application/json',
+  });
+};
+
+const timeout = (promise: Promise<Response>, ms = DEFAULT_TIMEOUT): Promise<Response> => {
+  const timerPromise = new Promise<never>((_, reject) => {
+    setTimeout(() => {
+      reject(new Error('Timeout'));
+    }, ms);
+  });
+  return Promise.race([timerPromise, promise]);
+};
+
+const call = <T>(method: Method, path: string, data: unknown): Promise<T | void> => {
+  const url = `${backendUrl}${path}`;
+  const options = {
+    method,
+    headers: getHeaders(),
+    body: data ? JSON.stringify(data) : null,
+  };
+  return timeout(fetch(url, options)).then(response => {
+    if (!response.ok) {
+      throw new Error(`${response.status} ${response.body}`);
+    }
+    if (response.status === 204) {
+      return null;
+    }
+    return response.text().then(text => (text ? JSON.parse(text) : response));
+  });
+};
+
+const handleError = (error: Error, path: string, message: string): void => {
+  console.error(`${path} failed: ${error.message}`);
+  showNotification({ message, severity: 'error' });
+};
+
+export function doRequest<T>(method: Method, path: string, errorMessage: string, data: unknown): Promise<T | void>;
+export function doRequest<T, R>(
+  method: Method,
+  path: string,
+  errorMessage: string,
+  data: unknown,
+  isExpectedResponse: (responseData: T) => boolean,
+  onSuccess: (responseData: T) => R
+): Promise<R | void>;
+
+export function doRequest<T, R>(
+  method: Method,
+  path: string,
+  errorMessage = `rest.error.${method.toLowerCase()}`,
+  data: unknown,
+  isExpectedResponse?: (responseData: T | void) => boolean,
+  onSuccess?: (responseData: T) => R
+): Promise<T | R | void> {
+  return call<T>(method, path, data)
```

Comment:
The same  could be done here but you would need to have a try catch block to give the catch functionality. I don't know which one I like better tbh

Reply:
I'm still not 100% sold on the whole onSuccess thing and the handleError inside here, but I think as we migrate the other calls over it'll become clear whether or not this is the right approach and then we can cleanup at that point.